File: languages/php/Bloock/Entity/Identity/CredentialBuilder.php
<?php

namespace Bloock\Entity\Identity;

use Bloock\Bridge\Bridge;
use Bloock\ConfigData;
use Bloock\CreateCredentialRequest;
use Bloock\Key as KeyProto;
use DateTime;
use Exception;

/**
 * Helps construct credentials by specifying various attributes.
 */
class CredentialBuilder
{
    private string $schemaId;
    private string $holderDid;
    private int $expiration;
    private int $version;
    private KeyProto $key;
    private ConfigData $configData;

    private array $stringAttributes;
    private array $integerAttributes;
    private array $decimalAttributes;
    private array $booleanAttributes;
    private array $dateAttributes;
    private array $datetimeAttributes;

    /**
     * Creates a new CredentialBuilder instance with the specified parameters.
     * @param string $schemaId
     * @param string $issuerDid
     * @param string $holderDid
     * @param int $expiration
     * @param int $version
     * @param ConfigData $configData
     */
    public function __construct(Issuer $issuer, string $schemaId, string $holderDid, int $expiration, int $version, ConfigData $configData)
    {
        $this->schemaId = $schemaId;
        $this->holderDid = $holderDid;
        $this->expiration = $expiration;
        $this->version = $version;
        $this->key = $issuer->getKey()->toProto();
        $this->configData = $configData;

        $this->stringAttributes = [];
        $this->integerAttributes = [];
        $this->decimalAttributes = [];
        $this->booleanAttributes = [];
        $this->dateAttributes = [];
        $this->datetimeAttributes = [];
    }

    /**
     * Adds a string attribute to the CredentialBuilder.
     * @param string $key
     * @param string $value
     * @return $this
     */
    public function withStringAttribute(string $key, string $value): CredentialBuilder
    {
        $attribute = new StringAttribute($key, $value);
        $this->stringAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Adds an integer attribute to the CredentialBuilder.
     * @param string $key
     * @param int $value
     * @return $this
     */
    public function withIntegerAttribute(string $key, int $value): CredentialBuilder
    {
        $attribute = new IntegerAttribute($key, $value);
        $this->integerAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Adds a decimal attribute to the CredentialBuilder.
     * @param string $key
     * @param float $value
     * @return $this
     */
    public function withDecimalAttribute(string $key, float $value): CredentialBuilder
    {
        $attribute = new DecimalAttribute($key, $value);
        $this->decimalAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Adds a boolean attribute to the CredentialBuilder.
     * @param string $key
     * @param bool $value
     * @return $this
     */
    public function withBooleanAttribute(string $key, bool $value): CredentialBuilder
    {
        $attribute = new BooleanAttribute($key, $value);
        $this->booleanAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Adds a date attribute to the CredentialBuilder.
     * @param string $key
     * @param DateTime $value
     * @return $this
     */
    public function withDateAttribute(string $key, DateTime $value): CredentialBuilder
    {
        $formated = $value->format('Y-m-d');
        $attribute = new DateAttribute($key, $formated);
        $this->dateAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Adds a datetime attribute to the CredentialBuilder.
     * @param string $key
     * @param DateTime $value
     * @return $this
     */
    public function withDatetimeAttribute(string $key, DateTime $value): CredentialBuilder
    {
        $formated = $value->format('Y-m-d\TH:i:sP');
        $attribute = new DatetimeAttribute($key, $formated);
        $this->datetimeAttributes[] = $attribute->toProto();
        return $this;
    }

    /**
     * Creates and returns a Credential using the specified attributes.
     * @return CredentialReceipt
     * @throws Exception
     */
    public function build(): CredentialReceipt
    {
        $bridge = new Bridge();

        $req = new CreateCredentialRequest();
        $req->setConfigData($this->configData);
        $req->setSchemaId($this->schemaId);
        $req->setHolderDid($this->holderDid);
        $req->setExpiration($this->expiration);
        $req->setVersion($this->version);
        $req->setKey($this->key);
        $req->setstringAttributes($this->stringAttributes);
        $req->setIntegerAttributes($this->integerAttributes);
        $req->setDecimalAttributes($this->decimalAttributes);
        $req->setBooleanAttributes($this->booleanAttributes);
        $req->setDateAttributes($this->dateAttributes);
        $req->setDatetimeAttributes($this->datetimeAttributes);

        $res = $bridge->identity->CreateCredential($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return CredentialReceipt::fromProto($res->getCredentialReceipt());
    }
}

File: languages/python/bloock/entity/identity/credential_builder.py
from __future__ import annotations
import datetime

from bloock._bridge import BloockBridge
from bloock._bridge.proto.config_pb2 import ConfigData
from bloock._bridge.proto.identity_pb2 import CreateCredentialRequest
from bloock._bridge.proto.shared_pb2 import Error
from bloock.entity.identity.boolean_attribute import BooleanAttribute
from bloock.entity.identity.date_attribute import DateAttribute
from bloock.entity.identity.datetime_attribute import DatetimeAttribute
from bloock.entity.identity.issuer import Issuer
from bloock.entity.identity.string_attribute import StringAttribute
from bloock.entity.identity.integer_attribute import IntegerAttribute
from bloock.entity.identity.decimal_attribute import DecimalAttribute
from bloock.entity.identity.credential_receipt import CredentialReceipt


class CredentialBuilder:
    """
    Helps construct credentials by specifying various attributes.
    """
    def __init__(self, issuer: Issuer, schema_id: str, holder_did: str, expiration: int, version: int, config_data: ConfigData) -> None:
        """
        Creates a new CredentialBuilder instance with the specified parameters.
        :type config_data: object
        :type version: object
        :type expiration: object
        :type holder_did: object
        :type schema_id: object
        :type issuer: object
        :rtype: object
        """
        self.schema_id = schema_id
        self.holder_did = holder_did
        self.expiration = expiration
        self.version = version
        self.key = issuer.key.to_proto()
        self.config_data = config_data

        self.string_attributes = []
        self.integer_attributes = []
        self.decimal_attributes = []
        self.boolean_attributes = []
        self.date_attributes = []
        self.datetime_attributes = []

    def with_string_attribute(self, key: str, value: str) -> CredentialBuilder:
        """
        Adds a string attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        self.string_attributes.append(StringAttribute(key, value))
        return self

    def with_integer_attribute(self, key: str, value: int) -> CredentialBuilder:
        """
        Adds an integer attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        self.integer_attributes.append(IntegerAttribute(key, value))
        return self

    def with_decimal_attribute(self, key: str, value: float) -> CredentialBuilder:
        """
        Adds a decimal attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        self.decimal_attributes.append(DecimalAttribute(key, value))
        return self

    def with_boolean_attribute(self, key: str, value: bool) -> CredentialBuilder:
        """
        Adds a boolean attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        self.boolean_attributes.append(BooleanAttribute(key, value))
        return self

    def with_date_attribute(self, key: str, value: datetime.date) -> CredentialBuilder:
        """
        Adds a date attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        formatted_date = value.strftime("%Y-%m-%d")
        self.date_attributes.append(DateAttribute(key, formatted_date))
        return self

    def with_datetime_attribute(self, key: str, value: datetime.datetime) -> CredentialBuilder:
        """
        Adds a datetime attribute to the CredentialBuilder.
        :type value: object
        :type key: object
        :rtype: object
        """
        rfc3339_string = value.strftime("%Y-%m-%dT%H:%M:%SZ")
        self.datetime_attributes.append(DatetimeAttribute(key, rfc3339_string))
        return self

    def build(self) -> CredentialReceipt:
        """
        Creates and returns a Credential using the specified attributes.
        :rtype: object
        """
        bridge = BloockBridge()

        string_attributes = []
        for a in self.string_attributes:
            string_attributes.append(a.to_proto())

        integer_attributes = []
        for a in self.integer_attributes:
            integer_attributes.append(a.to_proto())

        decimal_attributes = []
        for a in self.decimal_attributes:
            decimal_attributes.append(a.to_proto())

        boolean_attributes = []
        for a in self.boolean_attributes:
            boolean_attributes.append(a.to_proto())

        date_attributes = []
        for a in self.date_attributes:
            date_attributes.append(a.to_proto())

        datetime_attributes = []
        for a in self.datetime_attributes:
            datetime_attributes.append(a.to_proto())

        req = CreateCredentialRequest(
            config_data=self.config_data,
            schema_id=self.schema_id,
            holder_did=self.holder_did,
            version=self.version,
            expiration=self.expiration,
            string_attributes=string_attributes,
            integer_attributes=integer_attributes,
            decimal_attributes=decimal_attributes,
            boolean_attributes=boolean_attributes,
            date_attributes=date_attributes,
            datetime_attributes=datetime_attributes,
            key=self.key,
        )

        res = bridge.identity().CreateCredential(req)
        if res.error != Error():
            raise Exception(res.error.message)

        return CredentialReceipt.from_proto(res.credential_receipt)
